Tools/CISupport/ews-build/steps.py

```diff
@@ -2255,6 +2269,305 @@ def hideStepIf(self, results, step):
+                elif status == 2:  # failure
+                    failed_checks.append(queue)
+                else:  # null
+                    missing_checks.append(queue)
+
+        passed_status_check = self.getProperty('passed_status_check')
+        failed_status_check = self.getProperty('failed_status_check')
+        pending_prs = self.getProperty('pending_prs')
+
+        if len(missing_checks):
+            pending_prs.append(pr_number)
+            self.setProperty('pending_prs', pending_prs)
+            yield self._addToLog('stdio', 'Required checks are not completed. Waiting until all checks are completed before relabelling PR.\n')
+            yield self._addToLog('stdio', f'Missing the following checks: {missing_checks}\n')
+            defer.returnValue(False)
+        elif len(failed_checks):
+            failed_status_check.append(pr_number)
+            self.setProperty('failed_status_check', failed_status_check)
+            yield self._addToLog('stdio', 'Failed status check. Done!\n')
+            yield self._addToLog('stdio', f'Merged blocked due to failure: {failed_tests}\n')
+            defer.returnValue(False)
+        else:
+            passed_status_check.append(pr_number)
+            self.setProperty('passed_status_check', passed_status_check)
+            yield self._addToLog('stdio', 'Passed status check. Done!\n')
+            defer.returnValue(True)
+
+
+class AddMergeLabelsToPRs(buildstep.BuildStep, GitHubMixin, AddToLogMixin):
+    name = 'add-merge-labels-to-prs'
+    flunkOnFailure = False
+    haltOnFailure = False
+
+    @defer.inlineCallbacks
+    def run(self):
+        steps_to_add = []
+        label_as_safe = self.getProperty('passed_status_check', '')
+        yield self._addToLog('stdio', f'PRs to merge: {label_as_safe}.\n')
+        label_as_blocked = self.getProperty('failed_status_check', '')
+        yield self._addToLog('stdio', f'PRs to block: {label_as_blocked}.\n')
+        pending_prs = self.getProperty('pending_prs', '')
+        yield self._addToLog('stdio', f'PRs with checks pending: {pending_prs}. No action taken.\n')
+
+        if len(label_as_safe):
+            steps_to_add.append(RemoveAndAddLabels(label_to_add=GitHubMixin.MERGE_QUEUE_LABEL, labels_to_remove=[GitHubMixin.SAFE_MERGE_QUEUE_LABEL]))
+        if len(label_as_blocked):
+            steps_to_add.append(RemoveAndAddLabels(label_to_add=GitHubMixin.BLOCKED_LABEL, labels_to_remove=[GitHubMixin.SAFE_MERGE_QUEUE_LABEL]))
+        self.build.addStepsAfterCurrentStep(steps_to_add)
+        defer.returnValue(SUCCESS)
+
+    def getResultSummary(self):
+        if self.results == SUCCESS:
+            return {'step': f"Started PR labelling process successfully"}
+        elif self.results == FAILURE:
+            return {'step': f"Failed to start PR labelling process"}
+        return buildstep.BuildStep.getResultSummary(self)
+
+
+class RemoveAndAddLabels(buildstep.BuildStep, GitHubMixin, AddToLogMixin):
```

Comment:
Should we be consolidating with `BlockPullRequest` and `RemoveLabelsFromPullRequest`? Like previous steps (and especially if we're going to consolidate), we should probably be parameterizing `name`